apiserver/facades/client/charms/repositories.go

```diff
@@ -151,6 +170,42 @@ func (c *chRepo) findBundleDownloadURL(curl *charm.URL, origin corecharm.Origin)
 	return selector.Locate(info, origin)
 }
 
+// sanitizeCharmOrigin attempts to ensure that any fields we receive from
+// the API are valid and juju knows about them.
+func sanitizeCharmOrigin(received, requested params.CharmOrigin) (params.CharmOrigin, error) {
+	// Platform is generally the problem at hand. We want to ensure if they
+	// send "all" back for Architecture, OS or Series that we either use the
+	// requested origin using that as the hint or we unset it from the requested
+	// origin.
+	result := received
+	if received.Architecture == "all" {
+		result.Architecture = requested.Architecture
+	}
+	if received.OS == "all" {
+		result.OS = requested.OS
+	}
+	if received.Series == "all" {
+		result.Series = requested.Series
+
+		os, err := series.GetOSFromSeries(requested.Series)
+		if err != nil {
+			return received, errors.Trace(err)
+		}
+		result.OS = os.String()
```

Comment:
why overwrite OS if done at line 187? os must be lower case, and should be in the received, or the call would have failed.